cwltool/workflow.py

```diff
@@ -436,7 +479,79 @@ def __init__(self, toolpath_object, **kwargs):
         self.steps = [WorkflowStep(step, n, **kwargs) for n, step in enumerate(self.tool.get("steps", []))]
         random.shuffle(self.steps)
 
-        # TODO: statically validate data links instead of doing it at runtime.
+        # statically validate data links instead of doing it at runtime.
+        workflow_inputs = self.tool["inputs"]
+        workflow_outputs = self.tool["outputs"]
+
+        step_inputs = []; step_outputs = []
+        for step in self.steps:
+            step_inputs.extend(step.tool["inputs"])
+            step_outputs.extend(step.tool["outputs"])
+
+        # source parameters: workflow_inputs and step_outputs
+        # sink parameters: step_inputs and workflow_outputs
+
+        # make a dictionary of source parameters, indexed by the "id" field
+        src_parms = workflow_inputs + step_outputs
+        src_dict = {}
+        for parm in src_parms:
+            src_dict[parm["id"]] = parm
+
+        SrcSink = namedtuple("SrcSink", ["src", "sink", "linkMerge"])
+
+        def check_all_types(sinks, sourceField):
+            # type: (List[Dict[Text, Any]], Text) -> Dict[Text, List[SrcSink]]
+            # sourceField is either "soure" or "outputSource"
+            validation = {"warning": [], "exception": []}
+            for sink in sinks:
+                if sourceField in sink:
+                    valueFrom = sink.get("valueFrom")
+                    if isinstance(sink[sourceField], list):
+                        srcs_of_sink = [src_dict[parm_id] for parm_id in sink[sourceField]]
+                        linkMerge = sink.get("linkMerge", "merge_nested")
+                    else:
+                        parm_id = sink[sourceField]
+                        srcs_of_sink = [src_dict[parm_id]]
+                        linkMerge = sink.get("linkMerge")
+                    for src in srcs_of_sink:
+                        if check_types(src["type"], sink["type"], linkMerge, valueFrom) == "warning":
+                            validation["warning"].append(SrcSink(src, sink, linkMerge))
+                        elif check_types(src["type"], sink["type"], linkMerge, valueFrom) == "exception":
+                            validation["exception"].append(SrcSink(src, sink, linkMerge))
+            return validation
+
+        warnings = check_all_types(step_inputs, "source")["warning"] + \
+                    check_all_types(workflow_outputs, "outputSource")["warning"]
+        exceptions = check_all_types(step_inputs, "source")["exception"] + \
+                    check_all_types(workflow_outputs, "outputSource")["exception"]
```

Comment:
Please only call `check_all_types(step_inputs, "source")` and  `check_all_types(step_inputs, "source")` once and assign the results to variables, then you can separate the warnings and exceptions.